Bluetooth: ascs: Drop ISO PDUs if ASE is not in streaming state

MariuszSkamra opened the pull request:
This fixes missing drop of ISO Data PDUs received in non-streaming state.
The server shall indicate first it's readiness to receive the ISO Data
by calling bt_bap_stream_start that triggers state transition from
Enabling to Streaming state.

This adds tests for receiving ISO Data PDU.
* test_recv_in_streaming_state - test whether received ISO Data PDU is
  sent to upper layers.
* test_recv_in_enabling_state - test whether received ISO Data PDu is
  dropped because upper layer is not ready to receive the data yet.

MariuszSkamra committed:
Bluetooth: ascs: Drop ISO PDUs if ASE is not in streaming state

This fixes missing drop of ISO Data PDUs received in non-streaming state.
The server shall indicate first it's readiness to receive the ISO Data
by calling bt_bap_stream_start that triggers state transition from
Enabling to Streaming state.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
tests: Bluetooth: ascs: Add ISO Data PDU recv tests

This adds tests for receiving ISO Data PDU.
* test_recv_in_streaming_state - test whether received ISO Data PDU is
  sent to upper layers.
* test_recv_in_enabling_state - test whether received ISO Data PDu is
  dropped because upper layer is not ready to receive the data yet.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -671,10 +671,13 @@ static void ascs_iso_recv(struct bt_iso_chan *chan,
+				bt_bap_ep_state_str(ep->status.state));
+		}
+
+		net_buf_unref(buf);
```
Is the `net_buf_unref` correct to use here? IIRC it is no, and if it is, it needs to be added to _all_ cases.

MariuszSkamra replied:
That's surprise :) I see that there is no unref in samples as well. I thought that stack passes the buffer ownership

Thalley replied:
The stack unrefs the buf in `bt_iso_recv` by calling `bt_conn_reset_rx_state(iso);`

Thalley commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -671,10 +671,13 @@ static void ascs_iso_recv(struct bt_iso_chan *chan,
-	    ep->status.state != BT_BAP_EP_STATE_STREAMING) {
-		LOG_DBG("ep %p is not in the streaming state: %s", ep,
-			bt_bap_ep_state_str(ep->status.state));
+	if (ep->status.state != BT_BAP_EP_STATE_STREAMING) {
+		if (IS_ENABLED(CONFIG_BT_BAP_DEBUG_STREAM_DATA)) {
```
The issue you are fixing here by splitting the `if` is also missing in the unicast client. Could you update that too please?

MariuszSkamra replied:
sure

MariuszSkamra committed:
Bluetooth: unicast_client: Drop ISO PDUs if ASE is not in streaming state

This fixes missing drop of ISO Data PDUs received in non-streaming state.
The client shall indicate first it's readiness to receive the ISO Data.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>
